br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -1009,33 +1010,164 @@ func (local *DupeController) ResolveDuplicateRows(ctx context.Context, tbl table
 
 	errLimiter := rate.NewLimiter(1, 1)
 	pool := utils.NewWorkerPool(uint(local.dupeConcurrency), "resolve duplicate rows")
-	err = local.errorMgr.ResolveAllConflictKeys(
-		ctx, tableName, pool,
-		func(ctx context.Context, handleRows [][2][]byte) error {
-			for {
-				err := local.deleteDuplicateRows(ctx, logger, handleRows, decoder, keyInTable)
-				if err == nil {
-					return nil
-				}
-				if types.ErrBadNumber.Equal(err) {
-					logger.Warn("delete duplicate rows encounter error", log.ShortError(err))
-					return common.ErrResolveDuplicateRows.Wrap(err).GenWithStackByArgs(tableName)
-				}
-				if log.IsContextCanceledError(err) {
-					return err
+
+	switch algorithm {
+	case config.DupeResAlgRemove:
+		err = local.errorMgr.RemoveAllConflictKeys(
+			ctx, tableName, pool,
+			func(ctx context.Context, handleRows [][2][]byte) error {
+				for {
+					err := local.deleteDuplicateRows(ctx, logger, handleRows, decoder, keyInTable)
+					if err == nil {
+						return nil
+					}
+					if types.ErrBadNumber.Equal(err) {
+						logger.Warn("delete duplicate rows encounter error", log.ShortError(err))
+						return common.ErrResolveDuplicateRows.Wrap(errors.Trace(err)).GenWithStackByArgs(tableName)
+					}
+					if log.IsContextCanceledError(err) {
+						return errors.Trace(err)
+					}
+					if !tikverror.IsErrWriteConflict(errors.Cause(err)) {
+						logger.Warn("delete duplicate rows encounter error", log.ShortError(errors.Trace(err)))
+					}
+					if err = errLimiter.Wait(ctx); err != nil {
+						return errors.Trace(err)
+					}
 				}
-				if !tikverror.IsErrWriteConflict(errors.Cause(err)) {
-					logger.Warn("delete duplicate rows encounter error", log.ShortError(err))
+			},
+		)
+	case config.DupeResAlgReplace:
+		err = local.errorMgr.ReplaceConflictKeys(
+			ctx, tbl, tableName, pool, decoder,
+			func(ctx context.Context, key []byte) ([]byte, error) {
+				for {
+					value, err := local.getLatestValue(ctx, logger, key)
+					if err == nil {
+						return value, nil
+					}
+					if types.ErrBadNumber.Equal(err) {
+						logger.Warn("get latest value by key encounters error", log.ShortError(errors.Trace(err)))
+						return nil, common.ErrResolveDuplicateRows.Wrap(errors.Trace(err)).GenWithStackByArgs(tableName)
+					}
+					if log.IsContextCanceledError(errors.Trace(err)) {
+						return nil, errors.Trace(err)
+					}
+					if err = errLimiter.Wait(ctx); err != nil {
+						return nil, errors.Trace(err)
+					}
```

Comment:
Here is a for-loop and if we are careless it will become a dead loop. In DupeResAlgRemove branch seems `local.deleteDuplicateRows` can be repeated for arbitrary times and it will success finally, but `local.getLatestValue` will always return error